components/tikv_util/src/read_pool.rs

```diff
@@ -0,0 +1,398 @@
+                };
+
+                pool.spawn(f)?;
+            }
+            ReadPoolHandle::Yatp {
+                remote,
+                running_tasks,
+                max_tasks,
+            } => {
+                let running_tasks = running_tasks.clone();
+                // Note that the running task number limit is not strict.
+                // If several tasks are spawned at the same time while the running task number
+                // is close to the limit, they may all pass this check and the number of running
+                // tasks may exceed the limit.
+                if running_tasks.get() as usize >= *max_tasks {
+                    return Err(ReadPoolError::UnifiedReadPoolFull);
+                }
+
+                let fixed_level = match priority {
+                    CommandPri::High => Some(0),
+                    CommandPri::Normal => None,
+                    CommandPri::Low => Some(2),
+                };
+                let extras = Extras::new_multilevel(task_id, fixed_level);
+                let task_cell = TaskCell::new(
+                    async move {
+                        running_tasks.inc();
+                        f.await;
+                        running_tasks.dec();
+                    },
+                    extras,
+                );
+                remote.spawn(task_cell);
+            }
+        }
+        Ok(())
+    }
+
+    pub fn spawn_handle<F, T>(
+        &self,
+        f: F,
+        priority: CommandPri,
+        task_id: u64,
+    ) -> impl Future<Output = Result<T, ReadPoolError>>
+    where
+        F: Future<Output = T> + Send + 'static,
+        T: Send + 'static,
+    {
+        let (tx, rx) = oneshot::channel::<T>();
+        let res = self.spawn(
+            async move {
+                let res = f.await;
+                let _ = tx.send(res);
+            },
+            priority,
+            task_id,
+        );
+        async move {
+            res?;
+            rx.map_err(ReadPoolError::from).await
+        }
+    }
+}
+
+impl From<Vec<FuturePool>> for ReadPool {
+    fn from(mut v: Vec<FuturePool>) -> ReadPool {
+        assert_eq!(v.len(), 3);
+        let read_pool_high = v.remove(2);
+        let read_pool_normal = v.remove(1);
+        let read_pool_low = v.remove(0);
+        ReadPool::FuturePools {
+            read_pool_high,
+            read_pool_normal,
+            read_pool_low,
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct ReadPoolRunner<T: PoolTicker> {
+    inner: FutureRunner,
+    ticker: T,
+    after_start: Option<Arc<dyn Fn() + Send + Sync>>,
+    before_stop: Option<Arc<dyn Fn() + Send + Sync>>,
+    before_pause: Option<Arc<dyn Fn() + Send + Sync>>,
+}
+
+impl<T: PoolTicker> Runner for ReadPoolRunner<T> {
+    type TaskCell = TaskCell;
+
+    fn start(&mut self, local: &mut Local<Self::TaskCell>) {
+        if let Some(f) = self.after_start.take() {
+            f();
+        }
+        self.inner.start(local);
```

Comment:
```suggestion
        self.inner.start(local);
        if let Some(f) = self.after_start.take() {
            f();
        }
```
Should call `after_start` after start.